R/fit_best.R

```diff
@@ -0,0 +1,140 @@
+#'   wflow_fit <- fit(wflow, data_set)
+#' }
+#'
+#' The data used for the fit are taken from the `splits` column. If the split
+#' column was from a validation split, the combined training and validation sets
+#' are used.
+#'
+#' In comparison to [last_fit()], that function requires a finalized model, fits
+#' the model on the data set by [rsample::initial_split()], and computes
+#' metrics from the test set.
+#' @examples
+
+#' library(recipes)
+#' library(rsample)
+#' library(parsnip)
+#' library(dplyr)
+#'
+#' data(meats, package = "modeldata")
+#' meats <- meats %>% select(-water, -fat)
+#'
+#' set.seed(1)
+#' meat_split <- initial_split(meats)
+#' meat_train <- training(meat_split)
+#' meat_test <- testing(meat_split)
+#'
+#' set.seed(2)
+#' meat_rs <- vfold_cv(meat_train, v = 10)
+#'
+#' pca_rec <-
+#'   recipe(protein ~ ., data = meat_train) %>%
+#'   step_pca(all_predictors(), num_comp = tune())
+#'
+#' knn_mod <- nearest_neighbor(neighbors = tune()) %>% set_mode("regression")
+#'
+#' ctrl <- control_grid(save_workflow = TRUE)
+#'
+#' knn_pca_res <-
+#'   tune_grid(knn_mod, pca_rec, resamples = meat_rs, grid = 10, control = ctrl)
+#'
+#' knn_fit <- fit_best(knn_pca_res, verbose = TRUE)
+#' @return A fitted workflow.
+#' @export
+fit_best <- function(x, ...) {
+  UseMethod("fit_best")
+}
+
+#' @export
+#' @rdname fit_best
+fit_best.default <- function(x, ...) {
+  cls <- paste0("'", class(x), "'", collapse = ", ")
+  msg <-
+    paste("There is no `fit_best()` method for an object with class(es):", cls)
+  rlang::abort(msg)
+}
+
+#' @export
+#' @rdname fit_best
+fit_best.tune_results <- function(x, metric = NULL, parameters = NULL, verbose = FALSE, ...) {
+  if (length(list(...))) {
+    cli::cli_abort(c("x" = "The `...` are not used by this function."))
+  }
+  wflow <- .get_tune_workflow(x)
+  if (is.null(wflow)) {
+    cli::cli_abort(c("x" = "The control option `save_workflow = TRUE` should be used when tuning."))
+  }
+
+  if (is.null(parameters)) {
+    if (is.null(metric)) {
+      metric <- .get_tune_metric_names(x)[1]
+    }
+    parameters <- select_best(x, metric = metric)
+    if (verbose) {
+      format_final_param(parameters, metric)
+    }
+  }
+
+  # ----------------------------------------------------------------------------
+
+  wflow <- finalize_workflow(wflow, parameters = parameters)
+
+  # ----------------------------------------------------------------------------
+
+  pset <- hardhat::extract_parameter_set_dials(wflow)
+  p <- nrow(pset)
+  if (p > 0) {
+    prms <- paste0("'", pset$id, "'", collapse = ", ")
+    msg <- paste("There {?is/are} {p} parameter{?s} still marked for tuning: ", prms)
+    cli::cli_abort(
+      msg
+    )
+  }
+
+  # ----------------------------------------------------------------------------
+
+  dat <- x$splits[[1]]$data
+  if (verbose) {
+    cli::cli_inform(paste("Fitting using", nrow(dat), "data points..."))
```

Comment:
Are the `...` necessary? it is not immediately obvious to me (non-native speaker) what this is signaling

Reply:
I read the ellipses as indicating that the action messaged about is in progress rather than completed, but worth considering if you tripped up on this! Will leave unresolved for Max.

Reply:
Also, before I was using cli was just using cat. I would cat the first message, ending in the three dots and then write a 'done' on the same line if everything succeeded with the model fit.

Just so you know why those were originally there